### PR TITLE
Support a tab structure

### LINKED ISSUE #196: Support a tab structure
As I work on iquip project, I thought about supporing a tab structure in qiwis.
If we open several experiment builders, the GUI looks as below.

But I think showing the builders as a tab looks more legible using [`QTabWidgeet`](https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QTabWidget.html).

Feel free to leave your opinion😃

### PR DESCRIPTION
**Pull request by BECATRUE**
This PR will close #196.

I supported a tab structure, not showing in a row.

image

### CODE REVIEW

**Comment by BECATRUE** [comment]
I will modify the unittest after implementing the central widget.

File: qiwis.py [hunk 1]
```diff
@@ -203,7 +203,14 @@ def addFrame(self, name: str, frame: QWidget, info: AppInfo):
             "bottom": Qt.BottomDockWidgetArea
         }.get(info.pos, Qt.LeftDockWidgetArea)
         if info.show:
-            self.mainWindow.addDockWidget(area, dockWidget)
+            areaDockWidgets = [
+                dockWidget_ for dockWidget_ in self.mainWindow.findChildren(QDockWidget)
```
Comment (kangz12345) [inline qiwis.py:207]: 👍

File: qiwis.py [hunk 2]
```diff
@@ -203,7 +203,14 @@ def addFrame(self, name: str, frame: QWidget, info: AppInfo):
             "bottom": Qt.BottomDockWidgetArea
         }.get(info.pos, Qt.LeftDockWidgetArea)
         if info.show:
-            self.mainWindow.addDockWidget(area, dockWidget)
+            areaDockWidgets = [
+                dockWidget_ for dockWidget_ in self.mainWindow.findChildren(QDockWidget)
+                if self.mainWindow.dockWidgetArea(dockWidget_) == area
+            ]
+            if areaDockWidgets:
+                self.mainWindow.tabifyDockWidget(areaDockWidgets[-1], dockWidget)
```
Comment (kangz12345) [inline qiwis.py:211]: I think `areaDockWidgets[-1]` does not guarantee the order of tabbed widgets.
How about using [`tabifiedDockWidgets()`](https://doc.qt.io/qt-5/qmainwindow.html#tabifiedDockWidgets) to guarantee an intuitive order, e.g., the added tab is always comes to the front, etc.

Comment (kangz12345) [reply qiwis.py:211]: If `areaDockWidgets` preserves the order of widget addition, it would be okay.

Comment (BECATRUE) [reply qiwis.py:211]: > If areaDockWidgets preserves the order of widget addition, it would be okay.

Umm.. I think it doesn't guarantee the order😂

Comment (BECATRUE) [reply qiwis.py:211]: Now, I realize that `tabifyDockWidget()` always adds the `second` widget on the top of the area of the `first` widget.
(Here, the top means the last widget.)

For test, I changed it into `areaDockWidgets[0]`, but nothing was changed.

Comment (kangz12345) [reply qiwis.py:211]: Oh, good observation! Then never mind.

**Comment by BECATRUE** [comment]
Then, is there something to modify? @kangz12345

**Review by Aijuh — DISMISSED** [review]
I understand it as giving tab structure in the mainWindow. Not showing all the widget in one frame. 
Even if it has unittest error, I think you can talk to @kangz12345 about it. I will first approve it.

**Comment by kangz12345** [comment]
> I will modify the unittest after implementing the central widget.

Is the unit test error expected?

**Comment by BECATRUE** [comment]
> Is the unit test error expected?

No😮 I realized it just before and tried to fix it, but only unittest for Python 3.11 was failed.

**Comment by BECATRUE** [comment]
I found out the soluting using `patch.object()`.

**Comment by kangz12345** [comment]
Why was it a problem only in 3.11? Do you know the exact cause?

**Comment by BECATRUE** [comment]
> Why was it a problem only in 3.11? Do you know the exact cause?

Oh, I resolved it! I don't know the exact reason, but it looks like I did the wrong mocking.

**Comment by BECATRUE** [comment]
Please check this! @kangz12345 @Aijuh

**Comment by kangz12345** [comment]
I have 2 questions:
1. When we create an app at e.g. bottom and there were several frames (several columns), which one is tabified?
2. When we destroy an app whose frame was at e.g. bottom and there were 2 frames, will the remaining 1 frame be in a tab or not?

**Comment by BECATRUE** [comment]
> When we create an app at e.g. bottom and there were several frames (several columns), which one is tabified?

In my simple test, a new frame is created in the frame with the latest added frame, because I used `tabifyDockWidget()` with the last element of `areaDockWidgets` as follows:
```python
self.mainWindow.tabifyDockWidget(areaDockWidgets[-1], dockWidget)
```

If I use the first element `areaDockWidgets[0]`, the frame is created in the frame with the first added frame as we expect.

**Comment by BECATRUE** [comment]
> When we destroy an app whose frame was at e.g. bottom and there were 2 frames, will the remaining 1 frame be in a tab or not?

The remaining one is not in a tab, but just a single docked widget.
It seems that PyQt doesn't support a tab with only a frame.

**Comment by kangz12345** [comment]
Thank you for the confirmation!

**Review by kangz12345 — APPROVED** [review]
LGTM

**Comment by BECATRUE** [comment]
Please re-check this PR! @Aijuh

**Review by Aijuh — APPROVED** [review]

**Pull request merged**